### DISCUSSION TITLE
Documentation + source code search tips

### DISCUSSION THREAD

**Discussion by dylankb** [cue-lang/cue#2302] [General] [answered]
I'm curious if folks have strategies + tips for searching documentation or source code related to the Cue API.

Example tip: I'm just playing around with Cue on the side but I've found that if I'm searching for something that's in a Cue Go package then I can use Go documentation tools. E.g. browsing the [list](https://pkg.go.dev/cuelang.org/go@v0.4.3/pkg/list) package. 

Example stuck: I would like to find documentation (or if that's not available source code) for the Cue `or` builtin(?) function. It does not seem to be possible to use go to definition for the implementation since there's no LSP, so what do people do in these situations? 

I know there's a documentation overhaul effort underway, but looking for advice on what people do today.

**Comment by slewiskelly** [comment, marked as answer]
As already mentioned, [Go Doc](https://pkg.go.dev/cuelang.org/go) is useful.

For navigating source code, I use [Sourcegraph](https://sourcegraph.com/github.com/cue-lang/cue). Sourcegraph have been indexing public repos for some time now, but it doesn't seem to be very well known.

GitHub code search has improved a lot with the new code search (currently in beta). However, I still find Sourcegraph to be a superior experience.

Sourcegraph is also obviously useful if you want to find how other people are using something ([CUE example](https://sourcegraph.com/search?q=context:global+lang:cue+content:list.FlattenN&patternType=standard&sm=1&groupBy=repo), [Go example](https://sourcegraph.com/search?q=context:global+lang:go+content:cuecontext.New&patternType=standard&sm=1&groupBy=repo)).

**Reply by dylankb** [reply]
Great links. I didn't solve my specific stuck trying to find references to `or` ("or" returns a lot of results) from searching briefly, but these techniques already helped me solve another issue so I'll come back to this. Thanks!

**Reply by slewiskelly** [reply]
[Here](https://sourcegraph.com/search?q=context:global+lang:cue+content:or%5C%28%5C%5B&patternType=regexp&sm=1&groupBy=repo) are some uses of the `or` builtin.

**Comment by myitcv** [comment]
Thanks to @slewiskelly for chipping in here.

Also to flag that the Language Guide we are working on for the new version of cuelang.org will become the canonical place for learning language features. That will of course be complemented by lots of other sources of information, but we want the language guide to be a referencing learning point for those who want to learn in that way.
